File: cracking_the_coding_interview/arrays_strings/palindrome_permutation.py
"""
Given a string, write a function to check if it
is a permutation of a palin­ drome. A palindrome
is a word or phrase that is the same forwards
and backwards. A permutation is a rearrangement
of letters. The palindrome does not need to be
limited to just dictionary words.
EXAMPLE
Input: Tact Coa
Output: True (permutations: "taco cat", "atco eta", etc.)
"""

from collections import Counter


def permu_pal(string):
    length = len(string)
    count = Counter(string)
    odd_flag = False

    for c in count:
        if count[c] % 2 == 0:
            continue

        if length % 2 == 0:
            return False

        if not odd_flag:
            odd_flag = True
        else:
            return False

    return True


print(permu_pal("abccba"))
print(permu_pal("abcba"))
print(permu_pal("abcfba"))
print(permu_pal("aaaaa"))
print(permu_pal("aaaaav"))
print(permu_pal("aaaaavv"))

File: cracking_the_coding_interview/arrays_strings/palindrome_permutation.cpp
/*
Given a string, write a function to check if it
is a permutation of a palin­ drome. A palindrome
is a word or phrase that is the same forwards
and backwards. A permutation is a rearrangement
of letters. The palindrome does not need to be
limited to just dictionary words.
EXAMPLE
Input: Tact Coa
Output: True (permutations: "taco cat", "atco eta", etc.)
*/

#include <iostream>
#include <map>
#include <string>

std::string permuPal(std::string str)
{
  std::map<char, int> counter;
  for (char c : str)
  {
    auto position = counter.find(c);
    if (position == counter.end())
    {
      // item not found
      counter.insert(std::pair<char, int>(c, 1));
    }
    else
    {
      // item exists
      position->second++;
    }
  }
  const int STR_LEN = str.length();
  bool odd_flag = false;
  for (auto itr = counter.begin(); itr != counter.end(); itr++)
  {
    if (itr->second % 2 == 0)
      continue;

    if (STR_LEN % 2 == 0)
      return "false";

    if (!odd_flag)
    {
      odd_flag = true;
    }
    else
    {
      return "false";
    }
  }
  return "true";
}

int main()
{
  std::cout << permuPal("hello") << std::endl;
  std::cout << permuPal("abcba") << std::endl;
  std::cout << permuPal("abcfba") << std::endl;
  std::cout << permuPal("aaaaav") << std::endl;
  std::cout << permuPal("aaaaavv") << std::endl;
  return 0;
}
